src/Storages/Statistics/CountMinSketchStatistics.cpp

```diff
@@ -0,0 +1,88 @@
+#include <Storages/Statistics/CountMinSketchStatistics.h>
+#include <DataTypes/DataTypeNullable.h>
+#include <DataTypes/DataTypeLowCardinality.h>
+#include <IO/ReadHelpers.h>
+#include <IO/WriteHelpers.h>
+
+#if USE_DATASKETCHES
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int ILLEGAL_STATISTICS;
+}
+
+static constexpr auto num_hashes = 8uz;
+static constexpr auto num_buckets = 2048uz;
+
+CountMinSketchStatistics::CountMinSketchStatistics(const SingleStatisticsDescription & stat_, DataTypePtr data_type_)
+    : IStatistics(stat_)
+    , sketch(num_hashes, num_buckets)
+    , data_type(data_type_)
+{
+}
+
+Float64 CountMinSketchStatistics::estimateEqual(const Field & value) const
+{
+    if (auto float_val = IStatistics::getFloat64(value))
+        return sketch.get_estimate(&float_val.value(), 8);
+    if (auto string_val = IStatistics::getString(value))
+        return sketch.get_estimate(string_val->data(), string_val->size());
+    UNREACHABLE();
+}
+
+void CountMinSketchStatistics::serialize(WriteBuffer & buf)
+{
+    Sketch::vector_bytes bytes = sketch.serialize();
+    writeIntBinary(static_cast<UInt64>(bytes.size()), buf);
+    buf.write(reinterpret_cast<const char *>(bytes.data()), bytes.size());
+}
+
+void CountMinSketchStatistics::deserialize(ReadBuffer & buf)
+{
+    UInt64 size;
+    readIntBinary(size, buf);
+
+    Sketch::vector_bytes bytes;
+    bytes.reserve(size);
+    buf.readStrict(reinterpret_cast<char *>(bytes.data()), size);
+
+    sketch = Sketch::deserialize(bytes.data(), size);
+}
+
+void CountMinSketchStatistics::update(const ColumnPtr & column)
+{
+    size_t size = column->size();
+    for (size_t i = 0; i < size; ++i)
+    {
+        Field f;
+        column->get(i, f);
+
+        if (f.isNull())
+            continue;
+
+        if (auto float_val = IStatistics::getFloat64(f))
```

Comment:
The motivation of the convertion is to keep consistence of update and estimation of value.  

For the estimation example you provide, the value of `a` is represented by `Field` whose internal storage type is `UInt64`, there is no need to cast it to Int16, we can just convert it to `Float64` for all numeric types, which is a convinent way.

Reply:
In your case we need to check **if the constant is larger than std::numeric_limits<Int64>** . Otherwise converting `Int64->Float64` will also cause loss

Reply:
hi, now the estimation is base on column data type but not only the constant value in the lastest commit.